File: CodingChallenges/CC_07_SolarSystemGenerator_py/Planet.py
# Daniel Shiffman - Translated  to Python by Ali Al Saqban
# http://codingtra.in
# http://patreon.com/codingtrain

# Solar System Generator (2D)
# https://youtu.be/l8SiJ-RmeHU

class Planet(object):
    def __init__(self, r, d, o):
        self.radius = r
        self.distance = d
        self.angle = random(TWO_PI)
        self.planets = None
        self.orbitspeed = o
    
    def orbit(self):
        self.angle += self.orbitspeed
        if(self.planets):
            for i in self.planets:
                i.orbit()
    
    def spawnMoons(self,total,level):
        self.planets = []
        for i in range(total):
            r = self.radius/(level*2)
            d = random(50, 150)
            o = random(-0.1, 0.1)
            
            self.planets.append(Planet(r,d/level,o))
            
            if(level < 3):
                num = int(random(0,4))
                self.planets[i].spawnMoons(num, level + 1)
    
    def show(self):
        pushMatrix()
        
        fill(255,100)
        rotate(self.angle)
        translate(self.distance, 0)
        
        ellipse(0,0, self.radius*2, self.radius*2)
        
        if(self.planets):
            for i in self.planets:
                i.show()
        popMatrix()
        
                              

File: CodingChallenges/CC_007_SolarSystemGenerator_p5.js/planet.js
class Planet {
    constructor(radius, distance, orbitspeed, angle) {
        this.radius = radius;
        this.distance = distance;
        this.orbitspeed = orbitspeed;
        this.angle = angle;
        this.planets = [];
    }


    orbit() {
        this.angle += this.orbitspeed;
        for (let i in this.planets) {
            this.planets[i].orbit();
        }
    }


    spawnMoons(total, level) {
        for (let i = 0; i < total; i++) {
            let r = this.radius/(level*2);
            let d = random(50, 150);
            let o = random(-0.1, 0.1);
            let a = random(TWO_PI);
            this.planets.push(new Planet(r, d/level, o, a));
            if (level < 3) {
                let num = Math.floor(random(0, 4));
                this.planets[i].spawnMoons(num, level+1);
            }
        }
    }


    show() {
        push();
        fill(255, 100);
        rotate(this.angle);
        translate(this.distance, 0);
        ellipse(0, 0, this.radius*2);
        for (let i in this.planets) {
            this.planets[i].show();
        }
        pop();
    }

}

